src/backtrace.rs

```diff
@@ -0,0 +1,110 @@
+// SPDX-FileCopyrightText: 2023 Rivos Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+/// Get an array of backtrace addresses.
+///
+/// This needs `force-frame-pointers` enabled.
+use alloc::fmt::{Display, Formatter, Result};
+use core::arch::asm;
+use core::mem::size_of;
+
+extern "C" {
+    static _stack_start: u8;
+    static _stack_end: u8;
+}
+
+#[derive(Copy, Clone)]
+pub enum BTReturnAddress {
+    ReturnAddress(u64),
+    InvalidFramePointer(u64),
+    InvalidReturnAddress(u64),
+}
+
+impl Display for BTReturnAddress {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
+        match self {
+            BTReturnAddress::ReturnAddress(addr) => {
+                writeln!(f, "0x{addr:x}")
+            }
+            BTReturnAddress::InvalidFramePointer(addr) => {
+                writeln!(f, "Invalid frame pointer: 0x{addr:x}")
+            }
+            BTReturnAddress::InvalidReturnAddress(addr) => {
+                writeln!(f, "Null address at frame pointer: 0x{addr:x}")
+            }
+        }
+    }
+}
+
+pub struct BackTrace {
+    fp: Option<u64>,
+    stack_start: u64,
+    stack_end: u64,
+    stack: &'static [u64],
+}
+
+impl BackTrace {
+    fn new(fp: u64) -> Self {
+        let stack_start = unsafe { core::ptr::addr_of!(_stack_start) as u64 };
+        let stack_end = unsafe { core::ptr::addr_of!(_stack_end) as u64 };
+        let stack = unsafe {
+            core::slice::from_raw_parts(
+                stack_start as *const u64,
+                (stack_end - stack_start) as usize,
+            )
+        };
+
+        Self {
+            fp: Some(fp),
+            stack_start,
+            stack_end,
+            stack,
+        }
+    }
+
+    fn next_address(&self, fp: u64) -> Option<(u64, u64)> {
+        // if offset is out of bounds get will return none, and the whole function
+        // will return none
+        let offset = ((fp - self.stack_start) / size_of::<u64>() as u64) as usize;
+        let address = *self.stack.get(offset - 1)?;
```

Comment:
The issue here is the unchecked subtract.

To illustrate the problem, here's a sample program that demonstrates what happens if offset is 0 for example (it's silly to prevent the compiler from detecting the attempted subtract from 0 at compile time):

```
fn get_offset(b : bool) -> usize {
    if b {
        0
     } else { 4 }
}

fn main() {
    let stack = &[0, 1, 2, 3];
    let address = stack.get(foo(true) - 2);
    println!("{:?}", address);
}
```

Reply:
Since this is really just 2 addresses (stack_start and stack_start + 8) I guaranteed alignment and check for the addresses. I think that is enough of a check to fix the problem.